pkg/controller/statefulset/stateful_set_control_test.go

```diff
@@ -2423,51 +2423,48 @@ type requestTracker struct {
 	err      error
 	after    int
 
-	parallelLock sync.Mutex
-	parallel     int
-	maxParallel  int
-
-	delay time.Duration
-}
-
-func (rt *requestTracker) errorReady() bool {
-	rt.Lock()
-	defer rt.Unlock()
-	return rt.err != nil && rt.requests >= rt.after
-}
-
-func (rt *requestTracker) inc() {
-	rt.parallelLock.Lock()
-	rt.parallel++
-	if rt.maxParallel < rt.parallel {
-		rt.maxParallel = rt.parallel
+	// this block should be updated consistently
+	parallelLock         sync.Mutex
+	parallelRequests     int
+	maxParallelRequests  int
+	parallelRequestDelay time.Duration
+}
+
+func (rt *requestTracker) trackParallelRequests() {
+	if rt.parallelLock.TryLock() {
+		// lock acquired: we are the only or the first concurrent request
+		// initialize the next set of parallel requests
+		rt.parallelRequests = 1
+	} else {
+		// lock is held by other requests
+		// now wait for the lock to increase the parallelRequests
+		rt.parallelLock.Lock()
+		rt.parallelRequests++
 	}
-	rt.parallelLock.Unlock()
-
-	rt.Lock()
-	defer rt.Unlock()
-	rt.requests++
-	if rt.delay != 0 {
-		time.Sleep(rt.delay)
+	defer rt.parallelLock.Unlock()
+	// update the local maximum of parallel collisions
+	if rt.maxParallelRequests < rt.parallelRequests {
+		rt.maxParallelRequests = rt.parallelRequests
+	}
+	// increase the chance of collisions
+	if rt.parallelRequestDelay != 0 {
+		time.Sleep(rt.parallelRequestDelay)
 	}
 }
 
-func (rt *requestTracker) reset() {
-	rt.parallelLock.Lock()
-	rt.parallel = 0
```

Comment:
Turning into `rt.parallelRequests`. So from now on `rt.parallelRequests` is never reset. So originally `rt.parallel` was counting the number of requests in parallel until an error was ready. Followed by resting it. Why is it no longer needed to reset it?

Reply:
https://github.com/kubernetes/kubernetes/pull/124624#discussion_r1592923857